### ISSUE TITLE
Database connection problem after upgrading to v2.9.x

### ISSUE THREAD

**Issue by jorfermo** [Icinga/icinga2#6482]
After upgrading to v2.9.1 icingaweb2 is not showing any updates. Icinga2 is working properly.

Seems to be a problem with IDO MySQL. The database is working fine if I connect manually

This is the error from the logs

```
[2018-07-23 12:53:33 +0200] critical/IdoMysqlConnection: Error "MySQL server has gone away" when executing query "DELETE FROM icinga_customvariables WHERE object_id = 79990;DELETE FROM icinga_customvariablestatus WHERE object_id = 79990;INSERT INTO icinga_customvariables (config_type, instance_id, is_json, object_id, varname, varvalue) VALUES ('1', 3, '0', 79990, 'geolocation', '39.4809269782,-0.3425534628');INSERT INTO icinga_customvariables (config_type, instance_id, is_json, object_id, varname, varvalue) VALUES ('1', 3, '0', 79990, 'host_type', 'VMWARE');INSERT INTO icinga_customvariables (config_type, instance_id, is_json, object_id, varname, varvalue) VALUES ('1', 3, '1', 79990, 'notification', '{\"mail\":{\"users\":[\"jorts\"]},\"period\":\"never\"}');INSERT INTO icinga_customvariables (config_type, instance_id, is_json, object_id, varname, varvalue) VALUES ('1', 3, '0', 79990, 'vmware_host', 'halley.cc.upv.es');INSERT INTO icinga_customvariables (config_type, instance_id, is_json, object_id, varname, varvalue) VALUES ('1', 3, '0', 79990, 'vmware_user', 'nagios@VSPHERE.LOCAL');UPDATE icinga_customvariablestatus SET instance_id = 3,  is_json = '0',  object_id = 79990,  status_update_time = FROM_UNIXTIME(1532343213),  varname = 'geolocation',  varvalue = '39.4809269782,-0.3425534628' WHERE object_id = 79990 AND varname = 'geolocation';UPDATE icinga_customvariablestatus SET instance_id = 3,  is_json = '0',  object_id = 79990,  status_update_time = FROM_UNIXTIME(1532343213),  varname = 'host_type',  varvalue = 'VMWARE' WHERE object_id = 79990 AND varname = 'host_type';UPDATE icinga_customvariablestatus SET instance_id = 3,  is_json = '1',  object_id = 79990,  status_update_time = FROM_UNIXTIME(1532343213),  varname = 'notification',  varvalue = '{\"mail\":{\"users\":[\"jorts\"]},\"period\":\"never\"}' WHERE object_id = 79990 AND varname = 'notification';UPDATE icinga_customvariablestatus SET instance_id = 3,  is_json = '0',  object_id = 79990,  status_update_time = FROM_UNIXTIME(1532343213),  varname = 'vmware_host',  varvalue = 'halley.cc.upv.es' WHERE object_id = 79990 AND varname = 'vmware_host';UPDATE icinga_customvariablestatus SET instance_id = 3,  is_json = '0',  object_id = 79990,  status_update_time = FROM_UNIXTIME(1532343213),  varname = 'vmware_user',  varvalue = 'nagios@VSPHERE.LOCAL' WHERE object_id = 79990 AND varname = 'vmware_user';UPDATE icinga_hosts SET action_url = '',  active_checks_enabled = '1',  address = '158.42.9.175',  address6 = '',  alias = 'editorial-ocr',  check_command_object_id = 315,  check_interval = '1',  check_timeperiod_object_id = 471,  config_hash = '36d6ce4ce195993560778a175bd2e8d7b7a373367a0b919d92bfed8175ac79a9',  config_type = '1',  display_name = 'editorial-ocr',  event_handler_enabled = '1',  flap_detection_enabled = '0',  freshness_checks_enabled = '1',  freshness_threshold = '60',  high_flap_threshold = '25',  host_object_id = 79990,  icon_image = '',  icon_image_alt = '',  instance_id = 3,  low_flap_threshold = '25',  max_check_attempts = '3',  notes = '',  notes_url = '',  notification_interval = '1440',  notifications_enabled = '1',  notify_on_down = '1',  notify_on_downtime = '1',  notify_on_flapping = '1',  notify_on_recovery = '1',  notify_on_unreachable = '1',  passive_checks_enabled = '1',  process_performance_data = '1',  retry_interval = '0.500000' WHERE host_object_id = 79990;DELETE FROM icinga_hostgroup_members WHERE host_object_id = 79990;INSERT INTO icinga_hostgroup_members (host_object_id, hostgroup_id, instance_id) VALUES (79990, 3695, 3);INSERT INTO icinga_hostgroup_members (host_object_id, hostgroup_id, instance_id) VALUES (79990, 1615, 3);DELETE FROM icinga_hostdependencies WHERE dependent_host_object_id = 79990;UPDATE icinga_hoststatus SET acknowledgement_type = '0',  active_checks_enabled = '1',  check_command = 'hostalive',  check_timeperiod_object_id = 471,  check_type = '0',  current_check_attempt = '1',  current_notification_number = '0',  current_state = '1',  endpoint_object_id = 3,  event_handler_enabled = '1',  flap_detection_enabled = '0',  has_been_checked = '0',  host_object_id = 79990,  instance_id = 3,  is_flapping = '0',  is_reachable = '1',  last_check = FROM_UNIXTIME(-1),  last_hard_state = '1',  last_hard_state_change = FROM_UNIXTIME(1532333855),  last_state_change = FROM_UNIXTIME(1532333855),  max_check_attempts = '3',  next_check = FROM_UNIXTIME(1532343241),  next_notification = FROM_UNIXTIME(1532420322),  normal_check_interval = '1',  notifications_enabled = '1',  original_attributes = 'null',  passive_checks_enabled = '1',  percent_state_change = '0',  problem_has_been_acknowledged = '0',  process_performance_data = '1',  retry_check_interval = '0.500000',  scheduled_downtime_depth = '0',  should_be_scheduled = '1',  state_type = '0',  status_update_time = FROM_UNIXTIME(1532343213) WHERE host_object_id = 79990"
[2018-07-23 12:53:33 +0200] critical/IdoMysqlConnection: Exception during database operation: Verify that your database is operational!
```

## Expected Behavior
Icingaweb2 showing updated statuts

## Current Behavior
Icingaweb2 shows a frozen status since the moment of the update

## Your Environment
* Icinga Web 2 version and modules (System - About):

boxydash | 0.0.1
doc | 2.6.0
grafana | master
map | 1.0.4
monitoring | 2.6.0
reporting | 0.0.0
statusmap | 0.0.0

* Version used (`icinga2 --version`): r2.9.1-1
* Operating System and version: Ubuntu 16.04
* Enabled features (`icinga2 feature list`): api checker command compatlog graphite ido-mysql livestatus mainlog notification
* Config validation (`icinga2 daemon -C`): information/cli: Finished validating the configuration file(s

**Comment by dnsmichi** [comment]
> Error "MySQL server has gone away"

would hit my alarm bells, and the first thing would be checking the health status of the MySQL server itself (logs, table integrity, connections, etc.). It may also be a matter of resources or OOM, that includes checking the entire system. Is MySQL running on the same host as Icinga?

**Comment by jorfermo** [comment]
Yes it's in the same box but it has plenty of resources. The mysql timeout settings are big enough and those queries failing in the log execute instantly if I do it manually.

I thought it was related to the bug found in 2.9.0 and it would be solved upgrading 2.9.1 but it didn't.

I'm observing that checks after upgrading to 2.9.1 seems to be updating in the web but the hosts that were marked down are not updating:

> Next check | in -1d 20h

They have negative times for next check and forcing a check from the web doesn't work

**Comment by dnsmichi** [comment]
Analyse your database performance and monitor its health. There's a couple of hints inside the docs for it. To me, this sounds like a performance problem, or a long last single transaction blocking DELETE statement on a table which is is huge.

https://www.icinga.com/docs/icinga2/latest/doc/08-advanced-topics/#monitoring-icinga-2

**Comment by jorfermo** [comment]
I managed to add IDO check and icinga says it's ok:

> ***** Service Monitoring on arecibo *****
> 
> ido-mysql on arecibo is OK!
> 
> Info:    Connected to the database server (Schema version: '1.14.3'). Queries per second: 502.267 Pending queries: 2.000.
> 
> When:    2018-07-25 10:04:09 +0200
> Service: ido-mysql
> Host:    arecibo

**Comment by dnsmichi** [comment]
That looks like a recovery notification mail, anything else on a historical level for the query throughput, possible bottlenecks, the overall health of your server? How's those Grafana graphs looking like?

Set aside, you can always query the REST API for specific object runtime state, let alone web interfaces.

**Comment by jorfermo** [comment]
Yes, I set the service when you told me few hours ago. We weren't checking ido so there's no history of the check.

The server has been running fine for years until we upgraded to 2.9.0. Is there a simple way to downgrade to 2.8.x to make sure that's not the source of the problem?

**Comment by jorfermo** [comment]
Ok. Downgraded to 2.8.4 and everything is working fine now. So I can confirm there's a problem with 2.9.1. This is the log of the boot after downgrading:

> [2018-07-25 17:43:55 +0200] information/Application: Shutting down...
> [2018-07-25 17:43:55 +0200] information/ApiListener: 'api' stopped.
> [2018-07-25 17:43:55 +0200] information/CheckerComponent: 'checker' stopped.
> [2018-07-25 17:43:55 +0200] information/CompatLogger: 'compatlog' stopped.
> [2018-07-25 17:43:55 +0200] information/ExternalCommandListener: 'command' stopped.
> [2018-07-25 17:46:47 +0200] information/DbConnection: 'ido-mysql' started.
> [2018-07-25 17:46:47 +0200] information/GraphiteWriter: 'graphite' started.
> [2018-07-25 17:46:47 +0200] information/LivestatusListener: 'livestatus' started.
> [2018-07-25 17:46:47 +0200] information/CheckerComponent: 'checker' started.
> [2018-07-25 17:46:47 +0200] information/NotificationComponent: 'notification' started.
> [2018-07-25 17:46:47 +0200] information/CompatLogger: 'compatlog' started.
> [2018-07-25 17:46:47 +0200] information/ExternalCommandListener: 'command' started.
> [2018-07-25 17:46:47 +0200] information/LivestatusListener: Created UNIX socket in '/run/icinga2/cmd/livestatus'.
> [2018-07-25 17:46:47 +0200] information/GraphiteWriter: Finished reconnecting to Graphite in 0.000174999 second(s).
> [2018-07-25 17:46:47 +0200] information/ConfigItem: Activated all objects.
> [2018-07-25 17:46:48 +0200] information/DbConnection: Resuming IDO connection: ido-mysql
> [2018-07-25 17:46:48 +0200] information/IdoMysqlConnection: 'ido-mysql' resumed.
> [2018-07-25 17:46:48 +0200] information/IdoMysqlConnection: MySQL IDO instance id: 3 (schema version: '1.14.3')
> [2018-07-25 17:46:53 +0200] information/WorkQueue: #5 (ApiListener, RelayQueue) items: 0, rate: 103.517/s (6211/min 6211/5min 6211/15min);
> [2018-07-25 17:46:53 +0200] information/WorkQueue: #6 (ApiListener, SyncQueue) items: 0, rate:  0/s (0/min 0/5min 0/15min);
> [2018-07-25 17:46:53 +0200] information/WorkQueue: #7 (GraphiteWriter, graphite) items: 0, rate: 0.0166667/s (1/min 1/5min 1/15min);
> [2018-07-25 17:46:53 +0200] information/WorkQueue: #8 (IdoMysqlConnection, ido-mysql) items: 4898, rate:  0/s (0/min 0/5min 0/15min); empty in 17737 days, 15 hours, 46 minutes and 53 seconds
> [2018-07-25 17:47:03 +0200] information/WorkQueue: #8 (IdoMysqlConnection, ido-mysql) items: 16566, rate:  0/s (0/min 0/5min 0/15min); empty in 14 seconds
> [2018-07-25 17:47:13 +0200] information/WorkQueue: #8 (IdoMysqlConnection, ido-mysql) items: 28676, rate:  0/s (0/min 0/5min 0/15min); empty in 23 seconds
> [2018-07-25 17:47:16 +0200] information/ExternalCommandListener: Executing external command: [1532533636] SCHEDULE_FORCED_SVC_CHECK;dbmysql1;APT;1532533636
> [2018-07-25 17:47:16 +0200] information/ExternalCommandListener: Executing external command: [1532533636] SCHEDULE_FORCED_SVC_CHECK;renablo;APT;1532533636
> [2018-07-25 17:47:16 +0200] information/ExternalCommandListener: Executing external command: [1532533636] SCHEDULE_FORCED_SVC_CHECK;elastic;APT;1532533636
> [2018-07-25 17:47:16 +0200] information/ExternalCommandListener: Executing external command: [1532533636] SCHEDULE_FORCED_SVC_CHECK;elastic3;APT;1532533636
> [2018-07-25 17:47:16 +0200] information/ExternalCommandListener: Executing external command: [1532533636] SCHEDULE_FORCED_SVC_CHECK;dbmysql2;APT;1532533636
> [2018-07-25 17:47:16 +0200] information/ExternalCommandListener: Executing external command: [1532533636] SCHEDULE_FORCED_SVC_CHECK;hen;APT;1532533636
> [2018-07-25 17:47:16 +0200] information/ExternalCommandListener: Executing external command: [1532533636] SCHEDULE_FORCED_SVC_CHECK;elastic2;APT;1532533636
> [2018-07-25 17:47:16 +0200] information/ExternalCommandListener: Executing external command: [1532533636] SCHEDULE_FORCED_SVC_CHECK;elasticapli;APT;1532533636
> [2018-07-25 17:47:16 +0200] information/ExternalCommandListener: Executing external command: [1532533636] SCHEDULE_FORCED_SVC_CHECK;renablo;CPU;1532533636
> [2018-07-25 17:47:16 +0200] information/ExternalCommandListener: Executing external command: [1532533636] SCHEDULE_FORCED_SVC_CHECK;renablo;Load;1532533636
> [2018-07-25 17:47:23 +0200] information/WorkQueue: #8 (IdoMysqlConnection, ido-mysql) items: 28568, rate: 212.5/s (12750/min 12750/5min 12750/15min); empty in infinite time, your task handler isn't able to keep up
> [2018-07-25 17:47:30 +0200] information/ExternalCommandListener: Executing external command: [1532533650] SCHEDULE_FORCED_SVC_CHECK;manzanito;API;1532533650
> [2018-07-25 17:47:30 +0200] warning/ExternalCommandListener: External command failed: Error: Cannot reschedule forced service check for non-existent service 'API' on host 'manzanito'
> [2018-07-25 17:47:33 +0200] information/WorkQueue: #8 (IdoMysqlConnection, ido-mysql) items: 121, rate: 714.333/s (42860/min 42860/5min 42860/15min);
> [2018-07-25 17:47:35 +0200] information/IdoMysqlConnection: Finished reconnecting to MySQL IDO database in 47.2572 second(s).
> [2018-07-25 17:47:43 +0200] information/WorkQueue: #8 (IdoMysqlConnection, ido-mysql) items: 6, rate: 736.633/s (44198/min 44198/5min 44198/15min);
> [2018-07-25 17:47:51 +0200] information/Application: Got reload command: Starting new instance.
> [2018-07-25 17:47:53 +0200] information/WorkQueue: #8 (IdoMysqlConnection, ido-mysql) items: 6, rate: 820.367/s (49222/min 49222/5min 49222/15min);
> [2018-07-25 17:47:03 +0200] information/WorkQueue: #8 (IdoMysqlConnection, ido-mysql) items: 16566, rate:  0/s (0/min 0/5min 0/15min); empty in 14 seconds
> [2018-07-25 17:47:13 +0200] information/WorkQueue: #8 (IdoMysqlConnection, ido-mysql) items: 28676, rate:  0/s (0/min 0/5min 0/15min); empty in 23 seconds

Let me know if I can do something to help debugging

**Comment by kevinstiller** [comment]
Same problem here with the newest updates.

**Comment by kevinstiller** [comment]
We had duplicate database entries in the comments. I had to purge everything in "icinga_comments" to get it back to work.

**Comment by kevinstiller** [comment]
After I changed a custom variable in a command definition via IcingaWeb2 from 92 to 88 i ran into duplicate entrys in my mariadb. I changed it back, and everything is OK again...


`[2018-08-08 09:39:21 +0200] critical/IdoMysqlConnection: Error "Duplicate entry '1-1208-2018-08-08 07:27:28-132' for key 'instance_id'" when executing query "DELETE FROM icinga_customvariables WHERE object_id = 319;DELETE FROM icinga_customvariablestatus WHERE object_id = 319;INSERT INTO icinga_customvariables (config_type, instance_id, is_json, object_id, varname, varvalue) VALUES ('1', 1, '0', 319, 'disk_kritisch', '97');INSERT INTO icinga_customvariables (config_type, instance_id, is_json, object_id, varname, varvalue) VALUES ('1', 1, '0', 319, 'disk_warnung', '88');INSERT INTO icinga_customvariables (config_type, instance_id, is_json, object_id, varname, varvalue) VALUES ('1', 1, '0', 319, 'ram_kritisch', '97');INSERT INTO icinga_customvariables (config_type, instance_id, is_json, object_id, varname, varvalue) VALUES ('1', 1, '0', 319, 'ram_warnung', '92');UPDATE icinga_customvariablestatus SET instance_id = 1,  is_json = '0',  object_id = 319,  status_update_time = FROM_UNIXTIME(1533713960),  varname = 'disk_kritisch',  varvalue = '97' WHERE object_id = 319 AND varname = 'disk_kritisch';UPDATE icinga_customvariablestatus SET instance_id = 1,  is_json = '0',  object_id = 319,  status_update_time = FROM_UNIXTIME(1533713960),  varname = 'disk_warnung',  varvalue = '88' WHERE object_id = 319 AND varname = 'disk_warnung';UPDATE icinga_customvariablestatus SET instance_id = 1,  is_json = '0',  object_id = 319,  status_update_time = FROM_UNIXTIME(1533713960),  varname = 'ram_kritisch',  varvalue = '97' WHERE object_id = 319 AND varname = 'ram_kritisch';UPDATE icinga_customvariablestatus SET instance_id = 1,  is_json = '0',  object_id = 319,  status_update_time = FROM_UNIXTIME(1533713960),`

**Comment by Crunsher** [comment]
I could not reproduce this issue. What I tried was upgrading from 2.8.4 to 2.9.1, my config was the default one with a bunch of comments and customvars.
One thing @kevinstiller when you say you changed a customvar through icingaweb, do you mean the director module?

**Comment by kevinstiller** [comment]
@Crunsher Yup.

Meanwhile I figured out, that the mainproblem was Nagstamon Acknowledges.

**Comment by jorfermo** [comment]
@kevinstiller How so? I'm also using nagstamon

**Comment by dnsmichi** [comment]
Needs more investigation, it is hard to reproduce or understand what's going on. Therefore removed from the 2.9.2 milestone.

**Comment by jorfermo** [comment]
@dnsmichi Any way I can provide more information? I had to roll back to 2.8.* but I can try the upgrade again if you give me some steps to debug it

**Comment by jorfermo** [comment]
I see this problem didn't make it into 2.10 either

**Comment by dnsmichi** [comment]
Apparently we have little to less resources to investigate and debug every issue. This is unfortunate, but this issue doesn't give me any clear shot where I could start debug.

**Comment by jorfermo** [comment]
As I stated before I have a box where I'm holding back any updates because it gets broken. I can take a snapshot and help you with debugging if needed. Just tell me how I can help.

**Comment by dbu1986** [comment]
@jorfermo  I have made similar observations at #6549, caused by very large plugin outputs. You can check if this case applies to your environment.

**Comment by jorfermo** [comment]
Unfortunately it's not the case

**Comment by dnsmichi** [comment]
2.9 unfortunately contained several bugs in the IDO feature which may have an influence here. I've fixed plenty of them for 2.11. If the issue still persists, please create a new one.
